### PR TITLE
Issue #1163 - Convert Sudoku to use templates

### PR DESCRIPTION
**Pull request by Mailkov**
-----
IA Page: https://duck.co/ia/view/sudoku

### CODE REVIEW

File: lib/DDG/Goodie/Sudoku.pm [hunk 1]
```diff
@@ -55,13 +39,40 @@ handle remainder => sub {
...
+    #switch 0 to more sensible placeholders
+	$str_output =~ s/0/_/g;
+    
+	return $str_output,
+    structured_answer => {
+        id => 'sudoku',
+        name => 'Answer',
+        data => {
+            title => 'Sudoku',
+            subtitle => 'Level: ' . $difficulty,
```
Comment (moollaza) [inline lib/DDG/Goodie/Sudoku.pm]: I think we should switch "Level:" for "Difficulty:"

File: lib/DDG/Goodie/Sudoku.pm [hunk 2]
```diff
@@ -55,13 +39,40 @@ handle remainder => sub {
...
+		push @sudoku_tables, [@number];
+    }    
+    
+    #switch 0 to more sensible placeholders
+	$str_output =~ s/0/_/g;
+    
+	return $str_output,
+    structured_answer => {
+        id => 'sudoku',
+        name => 'Answer',
```
Comment (moollaza) [inline lib/DDG/Goodie/Sudoku.pm]: We can probably go with "Games" here

File: share/goodie/sudoku/sudoku.css [hunk 3]
```diff
@@ -1,4 +1,4 @@
-.zci--answer table.sudoku td {
+.c-base__content table.sudoku td {
```
Comment (moollaza) [inline share/goodie/sudoku/sudoku.css]: I would say we should target it with `.zci--sudoku table.sudoku td` -- we always start the CSS rules with the class for the `zci--*` wrapper div as there's only 1 on the page.

File: t/Sudoku.t [hunk 4]
```diff
@@ -15,29 +15,95 @@ ddg_goodie_test(
 	],
 	"sudoku" => test_zci(
 		qr/^[0-9_].*[0-9_]$/s,
-		html => qr/.*\<table.*\<\/table\>.*/s,
+		structured_answer => {
```
Comment (moollaza) [inline t/Sudoku.t]: Looks like there's a mix of tabs + spaces. Let's switch it all to spaces :+1:

**Comment by moollaza** [comment]
This is now live on beta: https://beta.duckduckgo.com/?q=sudoku+easy&ia=games

@abeyang this one needs a little design help -- I think we should make everything a little bigger?

**Before:**

sudoku_easy_at_duckduckgo

---

**Here's what I came up with:**

sudoku_at_duckduckgo

- font-size `tx--21`
- input have transparent background, bottom border
- width of square is `2em` (from `1.5em`)

Thoughts?

**Comment by moollaza** [comment]
@Mailkov after some closer inspection, I'm worried that this library might not work. As well the Perl approach could cause problems, because the module states it `could take minutes` to generate a game.

We need to ensure it takes less than 1s, so the ideal approach may be a pure-JS game, where we can load the game once the page starts loading, instead of holding up the page rendering.

Some issues:

- after a few refreshes, the game no longer changes
- on harder difficulties, we're showing many missing tiles

---

Another benefit to the JS approach is that we can react to the game when it's solved, and other things, like reveal the solution.

If you'd like to continue working on this you're welcome to. If not, we can make an issue for someone else to investigate a JS approach -- I don't want to burden you with unexpected work :)

/cc @mintsoft I think you made this originally? Any thoughts/ideas?

**Comment by mintsoft** [comment]
@moollaza IIRC it was an abandoned PR; I think it probably makes sense to switch it out to JS if there already exists a generating library in JS (http://einaregilsson.com/sudoku-in-javascript/ looks plausible). Although it'd basically a new IA (I don't think this one has ever actually been live?) it might be worth just deleting this completely and starting afresh?

**Comment by Mailkov** [comment]
@moollaza @mintsoft 
about issues: 
> * after a few refreshes, the game no longer changes
> * on harder difficulties, we're showing many missing tiles

I think you aren't testing the last version ...

About JS approach , I completely agree ...
Some years ago, I have studied how to generate Sudoku for a Java program and its solution, so I can work on this, I could converting my old code from Java to Javascript or I could using the library suggested by @mintsoft  or other .... but not immediately, so if you want fix this issue quickly you are free to make an issue for someone else.

**Comment by moollaza** [comment]
@Mailkov @mintsoft I think the Pure JS version is ideal. 

>it might be worth just deleting this completely and starting afresh?

Sure. The design I proposed above still works, we just need to have the Goodie use JS.

@Mailkov if you'd like, you could adapt the lib @mintsoft mentioned: https://raw.githubusercontent.com/einaregilsson/sudoku.js/master/sudoku.js

It's about 700 LOC, but part of that handles the UI creation I think?

If we can find a small, actively maintained library I can make it available through `DDH.require()` and host in on DDG Servers outside of the repo.

**Comment by Mailkov** [comment]
@moollaza  @mintsoft
I'm not sure that the lib @mintsoft mentioned is the best way ...
Reading http://einaregilsson.com/sudoku-in-javascript/
I noticed this : 
>I also wanted to make sure that every generated puzzle could be solved using only logic. Humans are (surprisingly enough) not as good at backtracking search as computers, so whenever I have removed a square in the algorithm above I also check if the puzzle can be solved using only elimination techniques. If it can't I reject it. That also means that the puzzles generated are kind of simple, there are a number of techiques humans use in Sudoku that are not just simple elimination, but the puzzles I generate don't really require them. So, there is definitely room for improvement!

So the difficult level will be always simple. 
Usually, the difficult isn't determinated by number of empty cells!

and this:
>875 milliseconds per generated puzzle isn't exactly great. I was planning on generating a new puzzle at the start of every game on http://www.sudoku-webgame.com, but adding almost a second to the load time of the page simply wasn't good enough. I could always see the delay and it annoyed me. So, what I decided to do instead was to just pre-generate a good batch of puzzles and ship them with the game, and then randomly choose one when the game starts. It kind of annoys me, because I'd prefer to generate on-the-fly, but I want things to feel fast.

So this library have only a limitated number of Sudoku puzzles!

If you give me some weeks ... I can make a small, actively maintained library!

**Comment by moollaza** [comment]
@Mailkov thanks for pointing those out -- I agree that it looks like that lib won't be ideal

>If you give me some weeks ... I can make a small, actively maintained library!

That'd be amazing! There's no rush :)

**Pull request closed without merging**